### PR TITLE
Add FilmRate Mutation

### PR DESCRIPTION
**Pull request by xuorig**
@cjoudrey

### CODE REVIEW

File: app/controllers/graphql_controller.rb [hunk 1]
```diff
@@ -1,8 +1,13 @@
 class GraphqlController < ApplicationController
   def execute
+    context = {
```
Comment (xuorig) [inline app/controllers/graphql_controller.rb]: that was just for dev, we can remove once auth ships

Comment (cjoudrey) [reply app/controllers/graphql_controller.rb]: Cool, we can change now :D

File: app/models/graph/mutations/mutation.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Graph
+  module Mutations
+    Mutation = GraphQL::ObjectType.define do
+      name "Mutation"
+      field :movieRate, field: Graph::Mutations::FilmRate.field
```
Comment (cjoudrey) [inline app/models/graph/mutations/mutation.rb]: `:filmRate`.. or `:rateFilm`, but yeah maybe not `movie`.

Comment (xuorig) [reply app/models/graph/mutations/mutation.rb]: woops

File: app/controllers/graphql_controller.rb [hunk 3]
```diff
@@ -2,9 +2,14 @@ class GraphqlController < ApplicationController
   before_action :authenticate
 
   def execute
+    context = {
+      user: @user,
+    }
     query_string = params[:query].to_s
     variables = ensure_hash(params[:variables])
-    context = { user: @user }
+    context = {
```
Comment (xuorig) [inline app/controllers/graphql_controller.rb]: oops

File: app/models/graph/mutations/film_rate.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+      return_field :film, Graph::Types::Film
+      return_field :errors, !types[!Graph::Types::MutationError]
+
+      resolve ->(_, input, ctx) do
+        raise GraphQL::ExecutionError.new('Authentication required to rate a film.') unless user = ctx[:user]
+
+        film_id = Graph.parse_id(input['filmId'], Film)
+        film = Film.find_by(id: film_id) if film_id
+        raise GraphQL::ExecutionError.new('Invalid filmId.') unless film
```
Comment (cjoudrey) [inline app/models/graph/mutations/film_rate.rb:18]: @xuorig any objections? Feels like this is a dev error. 🤔

File: app/models/graph/schema.rb [hunk 5]
```diff
@@ -12,6 +13,8 @@ module Graph
 
     object_from_id ->(id, query_ctx) do
       gid = GlobalID.parse(id)
+      return unless gid
```
Comment (cjoudrey) [inline app/models/graph/schema.rb:16]: If `id` is garbage, `gid` will be `nil` and will 💥 later.

File: app/models/graph/mutations/film_rate.rb [hunk 6]
```diff
@@ -7,27 +7,27 @@ module Mutations
...
-        return unless film = Film.find_by(id: input["filmId"])
+        film_id = Graph.parse_id(input['filmId'], Film)
+        film = Film.find_by(id: film_id) if film_id
+        raise GraphQL::ExecutionError.new('Invalid filmId.') unless film
 
-        rating = user.ratings.build(
-          film: film,
-          rating: input["rating"]
-        )
+        rating = user.ratings.where(film: film).first_or_initialize
```
Comment (xuorig) [inline app/models/graph/mutations/film_rate.rb]: 🥇

File: app/models/rating.rb [hunk 7]
```diff
@@ -2,5 +2,5 @@ class Rating < ApplicationRecord
   belongs_to :user
   belongs_to :film
 
-  validates_inclusion_of :rating, in: 0..5
+  validates :rating, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 5 }
```
Comment (cjoudrey) [inline app/models/rating.rb:5]: Better error message when you do this.

File: app/models/graph/mutations/film_rate.rb (hunk 4 above)
Comment (xuorig) [reply app/models/graph/mutations/film_rate.rb:18]: hmm yeah, you're right. No real app would ask for an id as input i guess

File: app/models/graph/mutations/film_rate.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+module Graph
+  module Mutations
+    FilmRate = GraphQL::Relay::Mutation.define do
+      name "FilmRate"
+
+      input_field :filmId, !types.ID
+      input_field :rating, !types.Int
+
+      return_field :film, Graph::Types::Film
+      return_field :errors, !types[!Graph::Types::MutationError]
```
Comment (xuorig) [inline app/models/graph/mutations/film_rate.rb]: did you want to return rating too ? if you rebase you have a type

**Comment by xuorig** [comment]
🚢

**Pull request merged**